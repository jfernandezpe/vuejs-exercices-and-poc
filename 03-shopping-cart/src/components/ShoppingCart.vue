<template>
  <article>
    <ul>
      <li v-for="product in cartProducts" :key="product.id">
        {{ product.name }} {{ product.price }}
        <button class="removeFromCart" @click="removeFromCart(product.id)">
          Remove
        </button>
      </li>
    </ul>
    <p>Total price: {{ totalPrice }}</p>
    <button class="buy" @click="buy">Buy</button>
  </article>
</template>

<script>
export default {
  name: "ShoppingCart",
  props: {
    cartProducts: {
      type: Array,
      default: () => []
    }
  },
  computed: {
    totalPrice() {
      return this.cartProducts.reduce((total, { price }) => {
        return total + price;
      }, 0);
    }
  },
  methods: {
    removeFromCart(idToRemove) {
      this.$emit("removeFromCart", idToRemove);
    },
    buy() {
      this.$emit("buy");
    }
  }
};
</script>

<style lang="sass" scoped></style>
