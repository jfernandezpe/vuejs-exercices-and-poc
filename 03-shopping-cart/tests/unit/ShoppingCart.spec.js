import { shallowMount } from "@vue/test-utils";
import ShoppingCart from "@/components/ShoppingCart.vue";

describe("ShoppingCart.vue", () => {
  let wrapper;
  const iPhoneId = "02";
  beforeEach(() => {
    wrapper = shallowMount(ShoppingCart, {
      propsData: {
        cartProducts: [
          {
            id: "01",
            name: "Guitarra elÃ©ctrica",
            image: "https://picsum.photos/id/145/640/480",
            price: 1200.0
          },
          {
            id: iPhoneId,
            name: "iPhone",
            image: "https://picsum.photos/id/160/640/480",
            price: 1800.0
          }
        ]
      }
    });
  });
  it("should render the layout", () => {
    expect(wrapper.element).toMatchSnapshot();
  });
  it("should calculate the total price", () => {
    expect(wrapper.vm.totalPrice).toBe(3000);
  });
  describe("when a item remove button is clicked", () => {
    it("should trigger the event removeFromCart", () => {
      const iPhoneButton = wrapper.findAll(".removeFromCart").at(1);
      iPhoneButton.trigger("click");

      expect(wrapper.emitted().removeFromCart).toBeTruthy();
      expect(wrapper.emitted().removeFromCart[0]).toEqual([iPhoneId]);
    });
  });
  describe("when the buy button is clicked", () => {
    it("should trigger the event buy", async () => {
      wrapper.find(".buy").trigger("click");

      await wrapper.vm.$nextTick();
      expect(wrapper.emitted().buy).toBeTruthy();
    });
  });
});
