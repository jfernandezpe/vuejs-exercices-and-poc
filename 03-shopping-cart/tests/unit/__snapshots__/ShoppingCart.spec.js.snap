// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ShoppingCart.vue should render the layout 1`] = `
<article>
  <ul>
    <li>
      
      Guitarra el√©ctrica 1200 
      <button
        class="removeFromCart"
      >
        Remove
      </button>
    </li>
    <li>
      
      iPhone 1800 
      <button
        class="removeFromCart"
      >
        Remove
      </button>
    </li>
  </ul>
   
  <p>
    Total price: 3000
  </p>
   
  <button
    class="buy"
  >
    Buy
  </button>
</article>
`;
